vertical-pod-autoscaler/recommender/util/decaying_histogram.go

```diff
@@ -0,0 +1,126 @@
+// is multiplied by the factor of 2^((sampleTime - decayStart) / halfLife).
+// This means that the sample loses half of its weight ("importance") with
+// each halfLife period.
+// Since only relative (and not absolute) weights of samples matter, the
+// decayStart can be shifted at any time, which is equivalent to multiplying all
+// weights by a constant. In practice the decayStart is shifted forward whenever
+// the exponents become too large, to avoid floating point arithmetics overflow.
+type decayingHistogram struct {
+	histogram
+	// Decay half life period.
+	halfLife time.Duration
+	// Reference time for determining the relative age of samples.
+	// It is always an integer multiple of halfLife.
+	decayStart time.Time
+}
+
+// NewDecayingHistogram returns a new DecayingHistogram instance using given options.
+func NewDecayingHistogram(options *HistogramOptions, halfLife time.Duration) Histogram {
+	return &decayingHistogram{
+		histogram:  *NewHistogram(options).(*histogram),
+		halfLife:   halfLife,
+		decayStart: time.Time{},
+	}
+}
+
+func (h *decayingHistogram) Percentile(percentile float64) float64 {
+	return h.histogram.Percentile(percentile)
+}
+
+func (h *decayingHistogram) AddSample(value float64, weight float64, time time.Time) {
+	h.histogram.AddSample(value, weight*h.decayFactor(time), time)
+}
+
+func (h *decayingHistogram) SubtractSample(value float64, weight float64, time time.Time) {
+	h.histogram.SubtractSample(value, weight*h.decayFactor(time), time)
+}
+
+func (h *decayingHistogram) Merge(other Histogram) {
+	o := other.(*decayingHistogram)
+	if h.halfLife != o.halfLife {
+		panic("can't merge decaying histograms with different half life periods")
+	}
+	// Align the older decayStart with the younger one.
+	if h.decayStart.Before(o.decayStart) {
+		h.shiftDecayStart(o.decayStart)
+	} else if o.decayStart.Before(h.decayStart) {
+		o.shiftDecayStart(h.decayStart)
+	}
+	h.histogram.Merge(&o.histogram)
+}
+
+func (h *decayingHistogram) Equals(other Histogram) bool {
+	h2, typesMatch := (other).(*decayingHistogram)
+	return typesMatch && h.halfLife == h2.halfLife && h.decayStart == h2.decayStart && h.histogram.Equals(&h2.histogram)
+}
+
+func (h *decayingHistogram) IsEmpty() bool {
+	return h.histogram.IsEmpty()
+}
+
+func (h *decayingHistogram) String() string {
+	return fmt.Sprintf("decayStart: %v, halfLife: %v\n%s", h.decayStart, h.halfLife, h.histogram.String())
+}
+
+func (h *decayingHistogram) shiftDecayStart(newDecayStart time.Time) {
+	// Make sure the decay start is an integer multiple of halfLife.
+	newDecayStart = newDecayStart.Round(h.halfLife)
+	exponent := round(float64(h.decayStart.Sub(newDecayStart)) / float64(h.halfLife))
+	for bucket := h.histogram.minBucket; bucket <= h.histogram.maxBucket; bucket++ {
+		// Bucket weight *= 2^exponent.
+		h.histogram.bucketWeight[bucket] = math.Ldexp(h.histogram.bucketWeight[bucket], exponent)
+	}
+	h.histogram.totalWeight = math.Ldexp(h.histogram.totalWeight, exponent)
```

Comment:
I would probably make this a method in histogram that multiplies all buckets and totalWeight by the same constant. I might be thinking c++ too much, where things are private between classes.

Reply:
Done.